swift/ql/lib/codeql/swift/frameworks/StandardLibrary/String.qll

```diff
@@ -10,3 +12,15 @@ private class StringSource extends SourceModelCsv {
       ]
   }
 }
+
+/**
+ * A content implying that, if a `String` is tainted, then all its fields are tainted.
+ */
+private class StringFieldsInheritTaint extends TaintInheritingContent,
+  DataFlow::Content::FieldContent {
+  StringFieldsInheritTaint() {
+    this.getField().getEnclosingDecl().(ClassOrStructDecl).getFullName() = "String" or
+    this.getField().getEnclosingDecl().(ExtensionDecl).getExtendedTypeDecl().getFullName() =
+      "String"
+  }
+}
```

Comment:
This appears to be a sensible approach, but I will note a few rough edges in the case of `String`: `String.count`, `String.isEmpty`, `String.first` and `String.last`.  These things can't contain a lot of data (especially `String.isEmpty` which is a `Bool`) so it would be very difficult for an attacker to utilize a theoretical taint path through any of them.  In CPP in particular we've found taint flow through `strlen` (equivalent to `String.count`) usually produces many more false positive results than useful ones.

I'm not sure what the right solution should be.  We could try to exclude the ones we don't want from this `TaintInheritingContent`, or switch to an explicit modelling approach.  We could potentially block all taint through 'small' types everywhere, solving the problem for good (but would this limit some uses of taint?).  Or we could stick to our principles that these *are* taint flows and expect queries to sanitize on an ad-hoc basis if they tend to find false positive results?

Reply:
On the one hand, taint flow through small types irrelevant for most "code injection" type queries. On the other hand, there was Heartbleed.

Reply:
Good point, there have been plenty of real bugs related to tainted integers (malicious image width/height bugs spring to mind as a second example) - and as you suggest our recent focus on code injection queries may be biasing my viewpoint.

I remain somewhat nervous of taint flow through `strlen` / `.count`.  In CPP we've seen a lot of code like this:
```
char *buffer = (char *)malloc(sizeof(char) * strlen(inputString));
```
which is usually a reasonable thing to do, even if `inputString` is tainted.  There were other similar patterns, and in the end we blocked taint through `strlen`.  This particular example ought to be uncommon in Swift, because you don't need to explicitly allocate like this when you're using a proper `String` class.  I suppose we could take a wait-and-see approach on whether flow through `.count` is actually a problem (or a positive) on Swift?

Reply:
It may also be worth asking other language teams about this. Experience from C/C++ might not apply very well to this judgment call, but experience from Java/C#/JavaScript/Python and related languages might be a better indication of which direction which should go for Swift.

Reply:
In Java, we _usually_ (we aren't 100% consistent) don't consider things like `length` to be propagating taint, because there are few opportunities in a memory-safe language where a tainted integer could be a problem. Although IMHO, I think we _should_ have those models as well, because some queries actually look for tainted integers. Other queries not interested in that (like code injection) can always add a type-based sanitizer that filters out uninteresting types. 

In Swift, where memory management _can_ be a thing, I think modeling these kind of methods makes even more sense. Maybe `String.count` should be a specific sanitizer for overflow-like vulnerabilities instead of not modeling it at all?

Reply:
If we opt for leaving `.count` et al as taint propagating steps, maybe we could have a library predicate like `isConversionToSmallType(DataFlow::Node)`, so that queries that don't want this behavior can sanitize it out without reinventing the wheel.

Reply:
I'd say we keep all those fields/members as potential sources of taint until proven otherwise. As folks have mentioned here, even integers can be sources of taint in some cases. I'm wondering though if there's a way to implement @d10c suggestion without the query writer doing too much work here.